wurst/network/Network.wurst

```diff
@@ -0,0 +1,506 @@
+	private static constant STRING_CACHE = InitGameCache("s.w3v")
+
+	private var currentState = NetworkState.PREPARING
+
+	// mission key to be used for this instance in all game caches
+	private var mkey = GamecacheKeys.getKey(this castTo int)
+
+	// data buffer to store data before syncing and to read after syncing
+	private var dataBuffer = new HashBuffer
+
+	// encoder to send strings over the network
+	private var stringEncoder = new StringEncoder
+
+	// gamecache buffers to sync data in rounds
+	private var gcIntBuffer     = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.INTEGER)
+	private var gcRealBuffer    = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.REAL)
+	private var gcBooleanBuffer = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.BOOLEAN)
+	private var gcStringBuffer  = new GamecacheBuffer(STRING_CACHE, mkey, BufferType.INTEGER)
+
+	// metadata container
+	private var metadataStore = new MetadataStore
+
+	// the player that we are receiving data from
+	private player sender
+
+	// metadata about the transmission
+	private var meta = metadata(0, 0, 0, 0, 0)
+	private var metaReceived = false
+
+	// these double as write counters for the sender and read counters for the receivers
+	private var currentInt       = 0
+	private var currentReal      = 0
+	private var currentBoolean   = 0
+	private var currentStringInt = 0
+	private var currentRound     = 0
+
+	// finish callback
+	private NetworkFinishedCallback finishCallback = (Buffer _buffer) -> error("Network: did not specify any callback function")
+
+	construct(player sender)
+		this.sender = sender
+
+	ondestroy
+		destroy gcIntBuffer
+		destroy gcRealBuffer
+		destroy gcBooleanBuffer
+		destroy gcStringBuffer
+		destroy stringEncoder
+		destroy dataBuffer
+		destroy metadataStore
+
+	function getData() returns HashBuffer
+		return dataBuffer
+
+	// sends info about the amount of data to be expected to each player
+	private function sendMetadata()
+		// we should only send the metadata once
+		if SAFETY_CHECKS_ENABLED and currentState != NetworkState.PREPARING
+			error("Network: trying to send duplicate metadata")
+
+		if localPlayer == sender
+			// push all strings into the string encoder
+			execute(() -> begin
+				while dataBuffer.hasString()
+					stringEncoder.pushString(dataBuffer.readString())
+			end)
+
+			// encode all strings into ints
+			execute(() -> begin
+				stringEncoder.encode()
+			end)
+
+		// lock the data buffer so that no players can read or write to it while
+		// the transmission hasn't finished
+		dataBuffer.setMode(BufferMode.LOCKED)
+
+		// get the amount of individual entities to be synced for each type
+		let intCount      = dataBuffer.getIntCount()
+		let realCount     = dataBuffer.getRealCount()
+		let booleanCount  = dataBuffer.getBooleanCount()
+		let asciiIntCount = stringEncoder.getIntCount()
+		
+		// calculate max length in each entity buffer
+		let maxLength = max(max(intCount, realCount), max(booleanCount, asciiIntCount))
```

Comment:
should probably make a max(x,y,z) for maths package

Reply:
Added min/max variants for up to 6 arguments in Maths package